### ISSUE TITLE
`GitIgnoreSpec` behaviors differ from git

### ISSUE THREAD

**Issue by Starry-OvO** [cpburnz/python-pathspec#81]
Here is the demo for `pathspec`:

```python
from pathlib import Path

import pathspec

exclude_lines=["*", "!libfoo", "!libfoo/**"]
exclude_spec = pathspec.GitIgnoreSpec.from_lines(exclude_lines)

print(exclude_spec.match_file(Path("./libfoo/__init__.py")))
```

In `pathspec==0.11.2` the result is `True`, which means `./libfoo/__init__.py` **is** excluded.

---

Another demo to check the behavior of `Git`:

Directory structure be like

```
demo-project
├── libfoo
│   └── __init__.py
└── .gitignore
```

Then check ignorance with command `git check-ignore -v ./libfoo/__init__.py`

Expect no output, which means `./libfoo/__init__.py` is **not** excluded.

---

I suppose this difference is caused by the `priority` between directory and file patterns.

https://github.com/cpburnz/python-pathspec/blob/878be226c5324a4c5470c2ff86034d27c0734d70/pathspec/gitignore.py#L124-L136

Including pattern `!/libfoo/**` is treated as a directory pattern and will be overridden by the excluding file pattern `*`, thus the `./libfoo/__init__.py` will get excluded.

p.s. I've googled this `priority` mechanism several hours, but still could not find out any documentation mentioning it.

---

Other platform info:
**pathspec** 0.11.2
**Git** 2.40.1.windows.1
**Windows** 11 22H2 (22621.2134)
**Python** 3.11.4

**Comment by cpburnz** [comment]
Thanks for the bug report.

The `priority` mechanism is an implementation detail meant to replicate how Git seems to prioritize directories over files. It's not something officially from the Git documentation.

**Comment by cpburnz** [comment]
This is now fixed in v0.12.0.
